### PR TITLE
Add proto serialization of DuckDB maps

### LINKED ISSUE #1785: MAP types are not supported on version 21.0
**Describe the bug**
We recently released support for nested types during profiling, however I am finding that MAPs do not have profiling insights or table outputs.

**To Reproduce**
1. start Rill
2. model `map([100, 5], [42, 43])`
3. see no table, no profiling

**Expected behavior**
I expect to see top k and tables for MAP types.

**Screenshots**
Internal video here: https://www.notion.so/rilldata/MAP-types-not-working-in-21-0-62a1e1638d524411ae7483e675328bb0

### PR DESCRIPTION
**Pull request by begelundmuller**
Closes #1785.

One problem here was that JSON object keys must be strings, but DuckDB maps can have any key type. To address that, I made it serialize the DuckDB map key to a JSON string. However, it won't support complex key types, like using other maps or arrays as map keys ([details here](https://github.com/marcboeker/go-duckdb/pull/74)).

Also refactored a few things:
- Added `TableHead` query type to cache table rows result
- Moved `pbutil` to `runtime/pkg`
- Removed `valToPB` due to duplication

### CODE REVIEW

File: runtime/pkg/pbutil/pbutil.go [hunk 1]
```diff
@@ -93,6 +101,49 @@ func ToStruct(v map[string]any) (*structpb.Struct, error) {
 	return x, nil
 }
 
+// ToStructCoerceKeys converts a map with non-string keys to a google.protobuf.Struct.
+// It attempts to coerce the keys to JSON strings.
+func ToStructCoerceKeys(v map[any]any) (*structpb.Struct, error) {
+	x := &structpb.Struct{Fields: make(map[string]*structpb.Value, len(v))}
+	for k1, v := range v {
+		// Coerce k1 to a string.
+		k2, ok := k1.(string) // If k1 is a string, we're good
```
Comment (egor-ryashin) [inline runtime/pkg/pbutil/pbutil.go:110]: these comments actually mirror the code, looks redundant

File: runtime/pkg/pbutil/pbutil.go [hunk 2]
```diff
@@ -93,6 +101,49 @@ func ToStruct(v map[string]any) (*structpb.Struct, error) {
 	return x, nil
 }
 
+// ToStructCoerceKeys converts a map with non-string keys to a google.protobuf.Struct.
+// It attempts to coerce the keys to JSON strings.
+func ToStructCoerceKeys(v map[any]any) (*structpb.Struct, error) {
+	x := &structpb.Struct{Fields: make(map[string]*structpb.Value, len(v))}
+	for k1, v := range v {
+		// Coerce k1 to a string.
```
Comment (egor-ryashin) [inline runtime/pkg/pbutil/pbutil.go:109]: a redundant comment

File: runtime/pkg/pbutil/pbutil.go [hunk 3]
```diff
@@ -93,6 +101,49 @@ func ToStruct(v map[string]any) (*structpb.Struct, error) {
...
+		k2, ok := k1.(string) // If k1 is a string, we're good
+		if !ok {
+			// k1 is not a string.
```
Comment (egor-ryashin) [inline runtime/pkg/pbutil/pbutil.go:112]: a redundant comment

File: runtime/pkg/pbutil/pbutil.go [hunk 4]
```diff
@@ -93,6 +101,49 @@ func ToStruct(v map[string]any) (*structpb.Struct, error) {
...
+			// First encode it using ToValue (to correctly coerce time, big numbers, etc.)
+			val, err := ToValue(k1)
+			if err != nil {
+				return nil, err
+			}
+			// Then encode it as JSON
```
Comment (egor-ryashin) [inline runtime/pkg/pbutil/pbutil.go:118]: a redundant comment

**Comment by egor-ryashin** [comment]
And there're no tests that checks the key conversion

File: runtime/pkg/pbutil/pbutil.go [hunk 5]
```diff
@@ -69,6 +69,14 @@ func ToValue(v any) (*structpb.Value, error) {
 			return nil, err
 		}
 		return structpb.NewStructValue(v2), nil
+	case duckdb.Map:
+		return ToValue(map[any]any(v))
+	case map[any]any:
+		v2, err := ToStructCoerceKeys(v)
+		if err != nil {
+			return nil, err
+		}
```
[the remark this answers is not in the document]
Comment (begelundmuller) [reply runtime/pkg/pbutil/pbutil.go:73]: Done

File: runtime/pkg/pbutil/pbutil.go [hunk 6]
```diff
@@ -93,6 +101,49 @@ func ToStruct(v map[string]any) (*structpb.Struct, error) {
...
+	for k1, v := range v {
+		// Coerce k1 to a string.
+		k2, ok := k1.(string) // If k1 is a string, we're good
+		if !ok {
+			// k1 is not a string.
+			// First encode it using ToValue (to correctly coerce time, big numbers, etc.)
+			val, err := ToValue(k1)
+			if err != nil {
+				return nil, err
+			}
```
[the remark this answers is not in the document]
Comment (begelundmuller) [reply runtime/pkg/pbutil/pbutil.go:120]: Done

File: runtime/pkg/pbutil/pbutil.go [hunk 7]
```diff
@@ -93,6 +101,49 @@ func ToStruct(v map[string]any) (*structpb.Struct, error) {
...
+			// Then encode it as JSON
+			data, err := val.MarshalJSON()
+			if err != nil {
+				return nil, err
+			}
```
[the remark this answers is not in the document]
Comment (begelundmuller) [reply runtime/pkg/pbutil/pbutil.go:125]: Done

**Review by egor-ryashin — APPROVED** [review]

**Pull request merged**